### ISSUE TITLE
Task 15: Text Improvements

### ISSUE THREAD

**Issue by paulushub** [ElinamLLC/SharpVectors#18]
Complete the support of vertical texts
Complete the support of text paths
Complete multi-position and multi-angle-rotation of characters.

#### This work item was migrated from CodePlex
CodePlex work item ID: '421'
Vote count: '1'

**Comment by paulushub** [comment]
[jonfrost@2011/02/24]
This would be a great feature - down the road I will want this.

**Comment by paulushub** [comment]
[SelormeyPaul@2011/02/24]
Jon, you forgot to vote for it!

**Comment by paulushub** [comment]
[UnknownUser@2011/02/24]

**Comment by paulushub** [comment]
[jonfrost@2011/02/24]
I have done some work in this area - I can take this on after I finish the other task.

**Comment by paulushub** [comment]
[UnknownUser@2013/02/14]

**Comment by Pat28** [comment]
Improve vertical text support, glyph-orientation-vertical has been replaced by CSS text-orientation.
writing-mode: tb is no vertical-rl etc...

https://www.w3.org/TR/SVG/text.html#GlyphOrientationVerticalProperty

<svg width="400" height="400">
  <rect x="50" y="50" width="50" height="50" style="fill:rgb(0,0,255);stroke-width:3;stroke:rgb(0,0,0)" />
  Sorry, your browser does not support inline SVG.  

<text x="100" y="110">Hallo wereld 日㌀㌃</text>

<text x="100" y="120" style="writing-mode: vertical-lr">Hello 24world 日㌀㌃</text>

<text x="140" y="120" style="writing-mode: vertical-rl; ">Hello 24 world 日㌀㌃</text>

<text x="180" y="120" style="writing-mode: vertical-rl; text-orientation: upright">Hello 日㌀㌃</text>

</svg>

**Comment by paulushub** [comment]
@Pat28 I am really preparing for this feature. There is a real work involved here since much has changed; I have just added the WOFF2 support, will need direct parsing of the the font files to support both vertical metrics and emojis. 
The main problem is I need to drop a feature in the current source; support for text rendering **without** converting the glyphs to path. This limits the current text rendering mode, but dropping it will break current applications, so I am planning to move these features **SharpVectors 2.x**.

**Comment by Pat28** [comment]
![vertical_sample](https://user-images.githubusercontent.com/2316171/135475089-ef54c9bf-5fcd-4730-b60e-b49d88bda996.jpg)

![afbeelding](https://user-images.githubusercontent.com/2316171/135474617-f7b693f1-fbc3-46d0-9614-a63bb95075f7.png)

**Comment by Pat28** [comment]
> 
> 
> @Pat28 I am really preparing for this feature. There is a real work involved here since much has changed; I have just added the WOFF2 support, will need direct parsing of the the font files to support both vertical metrics and emojis. The main problem is I need to drop a feature in the current source; support for text rendering **without** converting the glyphs to path. This limits the current text rendering mode, but dropping it will break current applications, so I am planning to move these features **SharpVectors 2.x**.

will you include rendering support for both GDI+ and WPF etc?

**Comment by paulushub** [comment]
> @Pat28 will you include rendering support for both GDI+ and WPF etc?

Yes, that is the plan. The GDI+ renderer uses only path geometry for text rendering, but limited by GDI+ font features support.
WPF has a better font support through `DirectWrite`, but lacking colored font or emoji (available on `DirectWrite for Windows 10+`, but not supported by WPF) and vertical text.
GDI+ has vertical text support, which works well for the vertical Japanese texts.

By directly parsing the font files, we can have a simple shaping framework for any graphics.

**Comment by Pat28** [comment]
Did you also have a look at SkiaSharp which is a lot faster on Windows in multi-threaded environments; it does not use GDI+ but its own in-memory rendering.
It seems to be using something called HarfBuzzSharp for rendering texts.

**Comment by paulushub** [comment]
> @Pat28 Did you also have a look at SkiaSharp which is a lot faster on Windows in multi-threaded environments;

Yes, I have look into it. The initial support for SVG in the Skia library is limited, but there is a recent effort to extend it. I still do not know how far they will go with this.
There is also, [Svg.Skia](https://github.com/wieslawsoltes/Svg.Skia) based on the [SVG.NET](https://github.com/svg-net/SVG) and [SkiaSharp.Extended](https://github.com/mono/SkiaSharp.Extended). But I have not used any of these and cannot tell their feature sets.

I was considering the `SkiaSharp` to enable uses in web development, since MS stopped allowing WPF for such applications, but may go first with `SharpDX` (Direct2D + DirectWrite is faster than Skia + HarfBuzz) for Windows/Server applications. 
`SkiaSharp + HarfBuzzSharp` is quite large, only those using Linux server may be interested.

Finally, I am following the progress on .NET 6.x and it graphics system to evaluate supports. Lots of work to do 🤣

**Comment by valhentai** [comment]
Hi, has there been any progress on the rendering of vertical text ?
I grabbed the 1.8.3 version to convert a Svg to a XAML drawing but the text with `text-orientation:upright` is still displayed horizontaly.
Do you have any idea of when it will be implemented ?
